contracts/libraries/WeightedOracleLibrary.sol

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity >=0.5.0 <0.8.0;
+
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+
+/// @title Weighted Oracle library
+/// @notice Provides functions to integrate with different tier oracles of the same V3 pair
+library WeightedOracleLibrary {
+
+    struct TimeWeightedObservation {
+        int24 timeWeightedAverageTick;
+        uint256 timeWeightedHarmonicMeanLiquidity;
+    }
+
+    /// @notice Fetches time-weighted observations across different tiers of the same Uniswap V3 pair
+    /// @param pools Addresses of different tiers of the same Uniswap V3 pair that we want to observe
+    /// @param period Number of seconds in the past to start calculating the time-weighted observations
+    /// @return observations An array of obervations that have been time-weighted from (block.timestamp - period) to block.timestamp
+    function consult(address[] calldata pools, uint32 period) internal view returns (TimeWeightedObservation[] memory observations) {
+        require(period != 0, 'BP');
+
+        observations = new TimeWeightedObservation[](pools.length);
+        uint256 periodX160 = uint256(period) * type(uint160).max;
+
+        uint32[] memory secondsAgos = new uint32[](2);
+        secondsAgos[0] = period;
+        secondsAgos[1] = 0;
+
+        for (uint256 i; i < pools.length; i++) {
+          (int56[] memory tickCumulatives, uint160[] memory secondsPerLiquidityCumulativeX128s) = IUniswapV3Pool(pools[i]).observe(secondsAgos);
+          int56 tickCumulativesDelta = tickCumulatives[1] - tickCumulatives[0];
+          uint160 secondsPerLiquidityCumulativesDelta = secondsPerLiquidityCumulativeX128s[1] - secondsPerLiquidityCumulativeX128s[0];
+
+          observations[i].timeWeightedAverageTick = int24(tickCumulativesDelta / period);
+          if (tickCumulativesDelta < 0 && (tickCumulativesDelta % period != 0)) observations[i].timeWeightedAverageTick--;
+
+          observations[i].timeWeightedHarmonicMeanLiquidity = periodX160 / (uint256(secondsPerLiquidityCumulativesDelta) << 32);
```

Comment:
why are we shifting the denominator left 32? can't we just use `periodX128 / secondsPerLiquidityCumulativesDelta` and get the exact same precision?

Reply:
That was my initial approach, but I changed it based on the link you provided
https://github.com/Uniswap/uniswap-v3-core/blob/b2c5555d696428c40c4b236069b3528b2317f3c1/contracts/test/OracleEchidnaTest.sol#L65-L68

But I'll change it back, no problem.

Reply:
oops i don't remember why i did that, never read the code i linked 😃 
there might be a reason, let me circle back

Reply:
I don't fully understand why using periodX128 without the `secondsPerLiquidityCumulativesDelta` shifting could lead to the number exceeding `type(uint128).max`. I agree that I don't want it to exceed it, so I'll go with whatever approach makes it happen. I just thought that this would also achieve it:
```solidity
observations[i].timeWeightedHarmonicAverageLiquidity = periodX128 / secondsPerLiquidityCumulativesDelta;
```

Reply:
if period is 1 second, periodX128 is `1<<128`, and if liquidity is `type(uint128).max`, `secondsPerLiquidityCumulativesDelta` increases by `1<<128 / type(uint128).max == 1`, so you end up with `1<<128 / 1` which is g.t. `type(uint128).max`

Reply:
Yeah, just run into that same problem in one of my tests haha

Will keep the shifting approach then!